### PR TITLE
Ignore bytes after end of IP packet in BroadcastRawUDPConn.ReadFrom

### PR DESCRIPTION
**Pull request by lprylli**
When reading raw packets from the network, it can happen that the raw
ethernet packet read has undefined bytes after the end of the ip
packet (either from the network or in some cases from the local
stack).

Those bytes should not be passed to the dhcp-receiver otherwise the
option parser which is picky about final padding byte will silently
discard the dhcp-reply.

Tested: ast2500 bmc reproducing the issue + existing go test for coverage.
Signed-off-by: Loic Prylli <lprylli@netflix.com>

### CODE REVIEW

**Comment by pmazzini** [comment]
@lprylli Thanks for your contribution.

The other alternative is to stop reading after the `optEnd` https://github.com/insomniacslk/dhcp/blob/master/dhcpv4/options.go#L158

However, the approach of this PR is probably better.

@hugelgupf What is your call?

File: dhcpv4/nclient4/conn_linux.go [hunk 1]
```diff
@@ -150,7 +150,10 @@ func (upc *BroadcastRawUDPConn) ReadFrom(b []byte) (int, net.Addr, error) {
 			IP:   net.IP(ipHdr.SourceAddress()),
 			Port: int(udpHdr.SourcePort()),
 		}
-		return copy(b, buf.ReadAll()), srcAddr, nil
+		// Extra padding after end of IP packet shoudl be ignored,
```
Comment (pmazzini) [inline dhcpv4/nclient4/conn_linux.go]: s/shoudl/should/

File: dhcpv4/nclient4/conn_linux.go [hunk 2]
```diff
@@ -150,7 +150,10 @@ func (upc *BroadcastRawUDPConn) ReadFrom(b []byte) (int, net.Addr, error) {
...
+		// if not dhcp option parsing will fail.
+		dhcpLen := int(ipHdr.PayloadLength()) - UDPMinimumSize
```
Comment (pmazzini) [inline dhcpv4/nclient4/conn_linux.go]: Maybe use `udpLen` here to be consistent with line 140.

**Comment by pmazzini** [comment]
I commented with a couple of NITs. The rest looks good to me. Thanks.

File: dhcpv4/nclient4/conn_linux.go (hunk 1 above)
Comment (lprylli) [reply dhcpv4/nclient4/conn_linux.go]: done

File: dhcpv4/nclient4/conn_linux.go (hunk 2 above)
Comment (lprylli) [reply dhcpv4/nclient4/conn_linux.go]: done, renamed ipLen to ipMaxHdrLen, and udpLen to udpHdrLen to avoid any confusion.

**Comment by pmazzini** [comment]
Thanks!

**Pull request merged**